common/beerocks/bcl/include/bcl/network/socket_refactorings.h

```diff
@@ -0,0 +1,795 @@
+
+    const struct sockaddr *sockaddr() const override
+    {
+        return reinterpret_cast<const struct sockaddr *>(&m_address);
+    }
+    const socklen_t &length() const override { return m_length; }
+
+private:
+    sockaddr_un m_address = {};
+    socklen_t m_length;
+};
+
+class InternetAddress : public SockAddr {
+public:
+    InternetAddress(uint16_t port, uint32_t address = INADDR_ANY)
+    {
+        m_address.sin_family      = AF_INET;
+        m_address.sin_addr.s_addr = address;
+        m_address.sin_port        = htons(port);
+
+        m_length = sizeof(m_address);
+    }
+
+    const struct sockaddr *sockaddr() const override
+    {
+        return reinterpret_cast<const struct sockaddr *>(&m_address);
+    }
+    const socklen_t &length() const override { return m_length; }
+
+private:
+    sockaddr_in m_address = {};
+    socklen_t m_length;
+};
+
+class LinkLevelAddress : public SockAddr {
+public:
+    LinkLevelAddress(uint32_t iface_index, const sMacAddr &dst_mac)
+    {
+        m_address.sll_family  = AF_PACKET;
+        m_address.sll_ifindex = iface_index;
+        m_address.sll_halen   = sizeof(sMacAddr);
+        std::copy_n(dst_mac.oct, sizeof(sMacAddr), m_address.sll_addr);
+
+        m_length = sizeof(m_address);
+    }
+
+    const struct sockaddr *sockaddr() const override
+    {
+        return reinterpret_cast<const struct sockaddr *>(&m_address);
+    }
+    const socklen_t &length() const override { return m_length; }
+
+private:
+    sockaddr_ll m_address = {};
+    socklen_t m_length;
+};
+
+class NetlinkAddress : public SockAddr {
+public:
+    NetlinkAddress(uint32_t groups)
+    {
+        // TODO
+        m_address.nl_family = AF_NETLINK;
+        m_address.nl_groups = groups;
+
+        m_length = sizeof(m_address);
+    }
+
+    const struct sockaddr *sockaddr() const override
+    {
+        return reinterpret_cast<const struct sockaddr *>(&m_address);
+    }
+    const socklen_t &length() const override { return m_length; }
+
+private:
+    sockaddr_nl m_address = {};
+    socklen_t m_length;
+};
+
+/**
+ * This class is a wrapper for the socket file descriptor obtained with the accept() system call.
+ */
+class ConnectedSocket : public SocketAbstractImpl {
+public:
+    explicit ConnectedSocket(int fd) : SocketAbstractImpl(fd) {}
+};
+
+class RawSocket : public SocketAbstractImpl {
+public:
+    explicit RawSocket(uint16_t protocol = ETH_P_ALL)
+        : SocketAbstractImpl(socket(AF_PACKET, SOCK_RAW, htons(protocol)))
+    {
+    }
+};
+
+class UdpSocket : public SocketAbstractImpl {
+public:
+    UdpSocket() : SocketAbstractImpl(socket(AF_INET, SOCK_DGRAM, IPPROTO_IP)) {}
+};
+
+class TcpSocket : public SocketAbstractImpl {
+public:
+    TcpSocket() : SocketAbstractImpl(socket(AF_INET, SOCK_STREAM, IPPROTO_IP)) {}
+};
+
+class UdsSocket : public SocketAbstractImpl {
+public:
+    explicit UdsSocket() : SocketAbstractImpl(socket(AF_UNIX, SOCK_STREAM, 0)) {}

```

Comment:
It's AF_UNIX, not AF_LINUX, so I updated the comment. AF_UNIX/SOCK_SEQPACKET is fully POSIX.

Reply:
I've never used SOCK_SEQPACKET and don't know if there's any implication or risk. I think we can easily use SOCK_STREAM and do not assume that message boundaries are preserved when using UDS. 
https://github.com/prplfoundation/prplMesh/pull/538/files#r354864834

Reply:
Well, SOCK_SEQPACKET does *exactly* what we need, so why bother with a lot of complexity, that we know has a few shortcomings anyway? It's just a matter of using the appropriate tool for the job.

That said, the complexity has already been implemented, so we could just as well continue to use it.

Reply:
IIRC we don't use TCP sockets now. But if we used them, we would have to implement SOCK_STREAM sockets anyway. In this case we could decide to use the same implementation for both TCP and UDS.
But since there is no rush to implement SOCK_STREAM, I see no reason not to use SOCK_SEQPACKET.